### PR TITLE
More accurate message types for `email` stubs

### PR DESCRIPTION
**Pull request by max-muoto**
Fix long-standing `email.parser` issues detailed in https://github.com/python/typeshed/issues/2417 and https://github.com/python/typeshed/issues/10762

This takes the approach recommended by @andersk and utilizies generic defaults to help us get accurate types accounting for the default arguments here without any overloads.

### CODE REVIEW

File: stdlib/email/__init__.pyi [hunk 1]
```diff
@@ -1,17 +1,24 @@
...
+
 # Definitions imported by multiple submodules in typeshed
 _ParamType: TypeAlias = str | tuple[str | None, str | None, str]  # noqa: Y047
 _ParamsType: TypeAlias = str | None | tuple[str, str | None, str]  # noqa: Y047
 
-def message_from_string(s: str, _class: Callable[[], Message] = ..., *, policy: Policy = ...) -> Message: ...
-def message_from_bytes(s: bytes | bytearray, _class: Callable[[], Message] = ..., *, policy: Policy = ...) -> Message: ...
-def message_from_file(fp: IO[str], _class: Callable[[], Message] = ..., *, policy: Policy = ...) -> Message: ...
-def message_from_binary_file(fp: IO[bytes], _class: Callable[[], Message] = ..., *, policy: Policy = ...) -> Message: ...
+def message_from_string(s: str, _class: Callable[[], _M] = ..., *, policy: Policy[_M] | Policy = policy.compat32) -> _M: ...
```
Comment (max-muoto) [inline stdlib/email/__init__.pyi:14]: Without unioning `Policy` (despite the default) I saw some MyPy issues.

**Comment by andersk** [comment]
`Policy.message_factory`, `Policy.__init__(message_factory)`, `Policy.handle_defect`, `Policy.register_defect` should use `_M` rather than `Message`.

**Comment by max-muoto** [comment]
> `Policy.message_factory`, `Policy.__init__(message_factory)`, `Policy.handle_defect`, `Policy.register_defect` should use `_M` rather than `Message`.

Sounds good, wasn't 100% sure there. Let me fix that.

**Comment by max-muoto** [comment]
> `Policy.message_factory`, `Policy.__init__(message_factory)`, `Policy.handle_defect`, `Policy.register_defect` should use `_M` rather than `Message`.

https://github.com/python/typeshed/pull/12104/commits/c8e29823cd24901a280afe7c98a6a7ee8a40f4d8

File: stdlib/email/__init__.pyi [hunk 2]
```diff
@@ -1,17 +1,21 @@
+# pyright: reportInvalidTypeVarUse=false
```
Comment (max-muoto) [inline stdlib/email/__init__.pyi:1]: See https://github.com/microsoft/pyright/discussions/8084 for some context on why we need this.

**Comment by andersk** [comment]
Oh and `EmailPolicy.__init__(message_factory)` should return `EmailMessage` rather than `Message`.

**Comment by max-muoto** [comment]
> Oh and `EmailPolicy.__init__(message_factory)` should return `EmailMessage` rather than `Message`.

https://github.com/python/typeshed/pull/12104/commits/85ed29568738dc642f913c0a8482ab22407aaf78

**Comment by max-muoto** [comment]
@srittau It seems like you looked into making `Policy` generic at some point? Do you maybe want to review here?

**Comment by max-muoto** [comment]
Sorry to bug, but just wanted to follow-up again @srittau, or if maybe someone else is able to take a look.

File: stdlib/email/_policybase.pyi [hunk 3]
```diff
@@ -3,20 +3,22 @@ from collections.abc import Callable
 from email.errors import MessageDefect
 from email.header import Header
 from email.message import Message
-from typing import Any
+from typing import Any, Generic, TypeVar
 from typing_extensions import Self
 
+_M = TypeVar("_M", bound=Message, default=Message)
```
Comment (srittau) [inline stdlib/email/_policybase.pyi:9]:
```suggestion
_M = TypeVar("_M", bound=Message[str, str], default=Message[str, str])
```

File: stdlib/email/_policybase.pyi [hunk 4]
```diff
@@ -41,7 +43,7 @@ class Policy(_PolicyBase, metaclass=ABCMeta):
     @abstractmethod
     def fold_binary(self, name: str, value: str) -> bytes: ...
 
-class Compat32(Policy):
+class Compat32(Policy[Message]):
```
Comment (srittau) [inline stdlib/email/_policybase.pyi:46]:
```suggestion
class Compat32(Policy[Message[str, str]]):
```

**Review by srittau — APPROVED** [review]
LGTM, a few optional suggestions below.

**Comment by max-muoto** [comment]
> LGTM, a few optional suggestions below.

Thanks, accepted all of your suggestions.

**Comment by srittau** [comment]
Looking at the new primer hit, I think we need to make `EmailMessage` generic, too: `class EmailMessage(MIMEPart[_HeaderRegistryT, _HeaderRegistryParamT]): ...`.

Although I have to admit, I'm a bit lost in how all these classes work together.

**Comment by max-muoto** [comment]
> Looking at the new primer hit, I think we need to make `EmailMessage` generic, too: `class EmailMessage(MIMEPart[_HeaderRegistryT, _HeaderRegistryParamT]): ...`.
> 
> Although I have to admit, I'm a bit lost in how all these classes work together.

I'll take another deeper look in the next day or two, since it's a been a while since I worked on this originally now haha.

**Comment by max-muoto** [comment]
Closing until I have time to take a look at this again.

**Pull request closed without merging**